sktime/transformations/series/detrend/_deseasonalize.py

```diff
@@ -482,47 +432,60 @@ def fit(self, Z, X=None):
         """
         self._is_fitted = False
         z = check_series(Z, enforce_univariate=True)
-        self._set_y_len(z)
-        self._set_y_index(z)
+        self._Z_index = Z.index
         sp = check_sp(self.sp)
 
-        if sp > 1:
-            # apply seasonal decomposition
-            _seasonalizer = _STL(
-                z.values,
-                period=sp,
-                seasonal=self.seasonal,
-                trend=self.trend,
-                low_pass=self.low_pass,
-                seasonal_deg=self.seasonal_deg,
-                trend_deg=self.trend_deg,
-                low_pass_deg=self.low_pass_deg,
-                robust=self.robust,
-                seasonal_jump=self.seasonal_jump,
-                trend_jump=self.trend_jump,
-                low_pass_jump=self.low_pass_jump,
-            ).fit()
-            self.stl_model = _seasonalizer
-            self.seasonal_ = self.stl_model.seasonal
-        else:
-            self.stl_model = None
-            self.seasonal_ = np.zeros_like(z.values)
+        # The statsmodels.tsa.seasonal.STL can only deal with sp >= 2
+        if sp < 2:
+            raise ValueError("sp must be positive integer >= 2")
+        self._stl = _STL(
+            z.values,
+            period=sp,
+            seasonal=self.seasonal,
+            trend=self.trend,
+            low_pass=self.low_pass,
+            seasonal_deg=self.seasonal_deg,
+            trend_deg=self.trend_deg,
+            low_pass_deg=self.low_pass_deg,
+            robust=self.robust,
+            seasonal_jump=self.seasonal_jump,
+            trend_jump=self.trend_jump,
+            low_pass_jump=self.low_pass_jump,
+        ).fit()
+
+        self.seasonal_ = pd.Series(self._stl.seasonal, index=Z.index)
+        self.resid_ = pd.Series(self._stl.resid, index=Z.index)
+        self.trend_ = pd.Series(self._stl.trend, index=Z.index)
 
-        self.seasonal_ = pd.Series(self.seasonal_, index=Z.index)
         self._is_fitted = True
         return self
 
     def _transform(self, y):
+        if not self._Z_index.equals(y.index):
+            raise NotImplementedError(
+                """
+                STLTransformer is only a descriptive trasnformer and
+                can only transform data that was given in fit().
+                Please use Deseasonalizer or Detrender."""
+            )
         return y - self.seasonal_
 
     def _inverse_transform(self, y):
-        seasonal_index = self._align_seasonal_index(y)
-        return y + self.seasonal_[seasonal_index]
```

Comment:
> Which behavior is changed in your opinion?

@aiwalter, @eyalshafran - By removing `_align_seasonal_index` here we won't be able to use the transformer as a deseasonaliser, which was possible in v.0.8.1. Am I missing something here?

Reply:
it was not possible before, it has just returned NaN for outsample transformations. At least this is what I got.

Reply:
you can still use it as deseasonalizer, but only for the data that was given in `fit()`

Reply:
We can have transform behave like the inverse_transform which would fix the example above. 

One other issue is that the statsmodels implementation does not work for data with missing data points or partially works for data with NaN values. I should have checked for that in the original code before using a numpy array to calculate STL.

Reply:
@aiwalter you're right, the following code snippet returns NaNs for the version in main. The strange thing is that it works fine as a deseasonalizer for the 0.8.1. release.

```python
import numpy as np
from matplotlib import pyplot as plt
from sktime.utils.plotting import plot_series
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.transformations.series.detrend import STLTransformer, Detrender
from sktime.forecasting.ets import AutoETS
from sklearn.pipeline import Pipeline
from sktime.datasets import load_airline

y = load_airline()

forecaster = TransformedTargetForecaster([
    ("stl", STLTransformer(sp=12)),
    ("detrend", Detrender()),
    ("ets", AutoETS(trend=None, seasonal=None, error="add"))
])
forecaster.fit(y[:-12])
pred = forecaster.predict(fh=np.arange(12))

plot_series(y, pred)
plt.show()
```

0.8.1 result: 
![image](https://user-images.githubusercontent.com/64217214/144768395-368f71d0-6479-4d7e-b748-9cb0337b747f.png)

Reply:
do you know what has changed?

Reply:
@aiwalter - sorry for taking ages to reply. Found the "issue". It's line 511, added in #1611: 
```python
self.seasonal_ = pd.Series(self.seasonal_, index=Z.index)
```

If you comment this out in the code that's in main, then the code snippet above works.